NAS-109316 / 21.04 / Add Custom Catalog / Nas 109316

AlexKarpov98 opened the pull request:
- Completed first tab ( Available Applications except new install applications logic )
- Completed second tab ( Installed Applications )

Ornias1993 commented:
Awesome work Alex...

Some input I came across when creating (and working on) a catalog:
- Ensure applications with the same name name in different catalogs can be installed (currently the install menu's conflict)
- Different label colors for the "Catalog name" label would be nice, to differentiate custom catalogs from official
- When working on this it might be interesting to take note of the fact how currently the space for an "Application name" is rather... small, for applications with long names.

you might or might not have already taken these talking points into account, but just wanted to be sure they are documented.

kmoore134 commented:
@AlexKarpov98  - This looks amazing!! Found a small bug when trying to edit a catalog:

![image](https://user-images.githubusercontent.com/1933088/112347376-2f860b00-8c9d-11eb-96f8-dc14900be91c.png)

AlexKarpov98 commented:
@kmoore134 Sure, is expected behavior as it is still in progress. As I have mentioned above, first 2 tabs are ready for now. :)

kmoore134 commented:
Ahh, didn't catch that before. Thanks and keep up this good work, looks really sharp.

sonicaj reviewed (commented):
It's looking nice, good work. I have following concerns:

1) Let's please reword `Catalogs` to `Available Applications`

![Screenshot 2021-03-24 at 11 13 36 PM](https://user-images.githubusercontent.com/17968138/112362802-a4d1fa00-8cf6-11eb-9baf-5dbaee649828.png)

2) Can we please add help text in the search bar from the designs ?

![Screenshot 2021-03-24 at 11 14 25 PM](https://user-images.githubusercontent.com/17968138/112362834-b0bdbc00-8cf6-11eb-8219-76910d63370d.png)

3) Can we please have the first letter of the catalog name/train uppercase and rest of it lowercase ?

![Screenshot 2021-03-24 at 11 15 03 PM](https://user-images.githubusercontent.com/17968138/112362908-c632e600-8cf6-11eb-9742-0d6ccdb5bc89.png)

4) Can we please default to selecting all catalogs and user can change this later ? Usually users would like to see applications available from all the catalog when they click on Available Applications.

![Screenshot 2021-03-24 at 11 15 50 PM](https://user-images.githubusercontent.com/17968138/112363024-e19df100-8cf6-11eb-9f18-7b4062cfdb3e.png)

5) Let's please removing the trailing colon

![Screenshot 2021-03-24 at 11 22 33 PM](https://user-images.githubusercontent.com/17968138/112363897-d3040980-8cf7-11eb-9616-dc7060817a44.png)

6) Also we are not showing a sort of a tool tip when we hover over Update Available, i think we decided that we should show the available version when we hover above it.

Ornias1993 commented:
Maybe a bit off topic: But is there a reason the case on the labels is forced? Aka: Can't we just have the user decide on the case and get rid of the translation code from all-caps ->start-with-capital-rest-lower-case

AlexKarpov98 commented:
> Maybe a bit off topic: But is there a reason the case on the labels is forced? Aka: Can't we just have the user decide on the case and get rid of the translation code from all-caps ->start-with-capital-rest-lower-case

If I understood correctly, we just do it for better look but @sonicaj , any other thoughts?

Ornias1993 commented:
@AlexKarpov98 As these are relatively important changes to the Apps UI: I tried out your changes, but I seem to hit a wall somewhere... Something it wrong with loading TrueCharts, but the Chart_validator has no idea what... but it throws these errors during refresh inside the GUI (regardless which branch of TrueCharts I use):

![image](https://user-images.githubusercontent.com/7613738/112731205-2f546c80-8f36-11eb-9534-dd8d8428a2a5.png)

This is the lost of catalogs, as you can see I tried multiple:
![image](https://user-images.githubusercontent.com/7613738/112731239-4c893b00-8f36-11eb-98a0-f9485058c186.png)

Apps installed prior to the new GUI are detected and show up perfectly fine though:

![image](https://user-images.githubusercontent.com/7613738/112731259-67f44600-8f36-11eb-93c1-127614a9e1d3.png)

Ornias1993 commented:
One nitpick on the design:
you might want to slightly change the color of installed apps on mouse-over.

Because it isn't 100% clear that clicking an app will open a modal.

Ornias1993 commented on src/app/pages/applications/catalog/catalog.component.ts:
```diff
@@ -70,44 +72,54 @@ export class CatalogComponent implements OnInit {
+    this.appService.getAllCatalogItems().subscribe(res => {
+      this.catalogNames = [];
+      this.catalogApps = [];
+      res.forEach(catalog => {
+        this.catalogNames.push(catalog.label);
+        catalog.preferred_trains.forEach(train => {
+          for (let i in catalog.trains[train]) {  
+            let item = catalog.trains.charts[i];
+            let versions = item.versions;
+            let latest, latestDetails;
```
This is the thing breaking all trains not calling "charts".
Fix is relatively easy:

```suggestion
      res.forEach(catalog => {
        this.catalogNames.push(catalog.label);
        catalog.preferred_trains.forEach(train => {
          for (let i in catalog.trains[train]) {  
            let item = catalog.trains[train][i];
            let versions = item.versions;
            let latest, latestDetails;
```

Ornias1993 commented:
Another bug:
If two catalogs have an app with the same name, they previously got the same install modal (the official modal, in case a custom catalog conflicts with the official one).
I hoped this would be solved now that trains and catalogs where properly implemented, but this is still a thing...

Ornias1993 commented on src/app/pages/applications/catalog/catalog.component.html:
```diff
@@ -3,29 +3,28 @@
           </div>
             <div class="content" fxLayout="column" fxLayoutAlign="center center">
-                <div class="catalog-title">
+                <div class="catalog-title" (click)="showSummaryDialog(item.name)">
                     <h3 textLimiter threshold="10">{{ item.name }}</h3>
```
This should fix most longer-named Apps being cut-off too soon:
```suggestion
                <div class="catalog-title" (click)="showSummaryDialog(item.name)">
                    <h3 textLimiter threshold="12">{{ item.name }}</h3>
```

I've manually tested this, with multiple resolutions...
While "Transmission" (12 characters wide) now displays correctly, it doesn't  cause any unexpected sideeffects on other App names

Ornias1993 commented:
@AlexKarpov98 I removed all my previous comments and fixed them myself in https://github.com/truenas/webui/pull/5275

That PR points towards this PR and should solve a lot of minor and major bugs and UX-issues that are still present in this PR.

I manually tested all of the new features using those bugfixes, including mixing and matching different Catalogs, Trains and Apps... Including all features (info modals, deleting, installing etcetcetc).

I hope this speeds up the work on this PR to get in without issue before codefreeze ^^

AlexKarpov98 commented:
> @AlexKarpov98 I removed all my previous comments and fixed them myself in #5277
> 
> That PR points towards this PR and should solve a lot of minor and major bugs and UX-issues that are still present in this PR.
> 
> I manually tested all of the new features using those bugfixes, including mixing and matching different Catalogs, Trains and Apps... Including all features (info modals, deleting, installing etcetcetc).
> 
> I hope this speeds up the work on this PR to get in without issue before codefreeze ^^

@Ornias1993 Great! merged.

AlexKarpov98 commented:
@sonicaj Fixed all issues you mentioned. Please check it.

Ornias1993 commented:
> > @AlexKarpov98 I removed all my previous comments and fixed them myself in #5277
> > That PR points towards this PR and should solve a lot of minor and major bugs and UX-issues that are still present in this PR.
> > I manually tested all of the new features using those bugfixes, including mixing and matching different Catalogs, Trains and Apps... Including all features (info modals, deleting, installing etcetcetc).
> > I hope this speeds up the work on this PR to get in without issue before codefreeze ^^
> 
> @Ornias1993 Great! merged.

Hope you liked it :)

Ornias1993 commented on src/app/core/classes/common-utils.ts:
```diff
@@ -39,5 +39,9 @@ export class CommonUtils {
+		return text.charAt(0).toUpperCase() + text.slice(1);
+	}
 }
   
```
GH doesn't detect this as a new line, might be ssome line-ending mixup happening here...

AlexKarpov98 replied:
Sorry, I am not sure I am following you. :)

Ornias1993 replied:
Look at "files changed" end-of-file-whiteline warning. 
It isn't detecting line 47, most often because Linux/Unix and Windows end-of-lines are mixed in the file.

AlexKarpov98 replied:
Perhaps you can't pull above changes?

Ornias1993 replied:
No I'm saying GH isn't detecting a linebreak correctly.
Look at "files changed" in github, notice it gives an "end of file whiteline" warning icon. 
This is often because mixed lineendings.

Ornias1993 replied:
![image](https://user-images.githubusercontent.com/7613738/112875483-88eaa180-90c4-11eb-82d6-cac40be7a4f4.png)

I'm just giving you a hint... This is getting awkwardly tedieus to make clear...

AlexKarpov98 replied:
Exactly...xD

sonicaj reviewed (commented):
Really nice job so far ;)

I have following concerns:

1) About available/installed applications, we are not rendering the markdown text correctly in version readme.
2) Can we reword `Active` to `Healthy` And `Not Active` to `Unhealthy` or `Not Healthy` in application summary dialog please ?
3) If a version is not healthy, we wouldn't be retrieving it's readme in this case as we can't reliably guarentee it's going to exist for the catalog. Can we in this case please show the error string ? ( the key is `healthy_error` )

I have not verified the actual installation of different versions etc. After you give me the go ahead on that, we can do that as well.

Ornias1993 commented:
@sonicaj
1. Agreed, good catch... I also noticed it and forgot to mention
2. I'm not 100% a fan of healthy (because it might be confusing with unhealthy deployments/containers for docker users) However, I'm not sure if there is a beter way of saying this. However: indeed active should at the very least be Healthy on the application summary dialog.
3. Agreed... 

> I have not verified the actual installation of different versions etc. After you give me the go ahead on that, we can do that as well.

Different versions are the only thing I haven't tested yet.
I've tested:
- Multiple Catalogs
- Multiple Catalogs with duplicate Apps
- Multiple Trains
- Multiple Catalogs with Multiple Trains each
- Multiple Catalogs with Multiple Trains with duplicate names each
- Multiple Catalogs from the same repository in a different branch
- Multiple Catalogs from the same repository in a different branch including above train tests.

Ornias1993 commented:
@AlexKarpov98 I tested the container tab.
Some notes:
- Clicking "Update Image" doesn't generate the expected popup while updating the image
- Updating doesn't actually do anything at all
- It would be nice to be able to select multiple containers to upgrade

However: Generally speaking it looks fine! :)

---

Screenshot for those interested:
![image](https://user-images.githubusercontent.com/7613738/113064897-35f21680-91b8-11eb-9ae1-5c156744b026.png)

AlexKarpov98 commented:
> @AlexKarpov98 I tested the container tab.
> Some notes:
> 
> * Clicking "Update Image" doesn't generate the expected popup while updating the image
> * Updating doesn't actually do anything at all
> * It would be nice to be able to select multiple containers to upgrade
> 
> However: Generally speaking it looks fine! :)
> 
> Screenshot for those interested:
> ![image](https://user-images.githubusercontent.com/7613738/113064897-35f21680-91b8-11eb-9ae1-5c156744b026.png)

Last tab is in progress, not completed yet.
So the functions like updating images won't work for now.

AlexKarpov98 commented:
> @sonicaj
> 
> 1. Agreed, good catch... I also noticed it and forgot to mention
> 2. I'm not 100% a fan of healthy (because it might be confusing with unhealthy deployments/containers for docker users) However, I'm not sure if there is a beter way of saying this. However: indeed active should at the very least be Healthy on the application summary dialog.
> 3. Agreed...
> 
> > I have not verified the actual installation of different versions etc. After you give me the go ahead on that, we can do that as well.
> 
> Different versions are the only thing I haven't tested yet.
> I've tested:
> 
> * Multiple Catalogs
> * Multiple Catalogs with duplicate Apps
> * Multiple Trains
> * Multiple Catalogs with Multiple Trains each
> * Multiple Catalogs with Multiple Trains with duplicate names each
> * Multiple Catalogs from the same repository in a different branch
> * Multiple Catalogs from the same repository in a different branch including above train tests.

@Ornias1993 
Can you please share me the sample charts like multiple catalogs, multiple versions including healthy and unhealthy so that with which I can test all possible cases?

Ornias1993 commented:
> > @sonicaj
> > 
> > 1. Agreed, good catch... I also noticed it and forgot to mention
> > 2. I'm not 100% a fan of healthy (because it might be confusing with unhealthy deployments/containers for docker users) However, I'm not sure if there is a beter way of saying this. However: indeed active should at the very least be Healthy on the application summary dialog.
> > 3. Agreed...
> > 
> > > I have not verified the actual installation of different versions etc. After you give me the go ahead on that, we can do that as well.
> > 
> > 
> > Different versions are the only thing I haven't tested yet.
> > I've tested:
> > 
> > * Multiple Catalogs
> > * Multiple Catalogs with duplicate Apps
> > * Multiple Trains
> > * Multiple Catalogs with Multiple Trains each
> > * Multiple Catalogs with Multiple Trains with duplicate names each
> > * Multiple Catalogs from the same repository in a different branch
> > * Multiple Catalogs from the same repository in a different branch including above train tests.
> 
> @Ornias1993
> Can you please share me the sample charts like multiple catalogs, multiple versions including healthy and unhealthy so that with which I can test all possible cases?

Well you can try the TrueCharts staging branch, it containers a "stable" branch instead of the default "charts" branch and contains one duplicate chart with Official (Plex)

But for my personal tests, those where "throw-away" branches, so I don't really have anything to share (saidly enough)

AlexKarpov98 commented:
@Ornias1993 It's okay and already got enough instructions from Waqar :)
Anyway thank you for your sharing.

Ornias1993 commented:
We might actually want to look into some seperated testing catalogs with different trains and duplicate Apps, for the automated testing eventually. But for now you do a lot of good work anyway :)

AlexKarpov98 commented:
> @AlexKarpov98 I tested the container tab.
> Some notes:
> 
> * Clicking "Update Image" doesn't generate the expected popup while updating the image
> * Updating doesn't actually do anything at all
> * It would be nice to be able to select multiple containers to upgrade
> 
> However: Generally speaking it looks fine! :)
> 
> Screenshot for those interested:
> ![image](https://user-images.githubusercontent.com/7613738/113064897-35f21680-91b8-11eb-9ae1-5c156744b026.png)

@Ornias1993 Regarding multiple containers, it is not supported on middleware for now afaik. :)

AlexKarpov98 commented:
@sonicaj @Ornias1993 
All functionalities and issues on comments are fixed. So it should work for all possible cases.
But one thing, `markdown`, it is missing and still needs some time to implement. Actually don't have any estimation of it for now.

Ornias1993 commented:
@AlexKarpov98 I think leaving the markdown for now is fair, as long as we clearly document it on Jira...
It's a minor grievence really and I think 99% of other grievences are solved with this PR :)

sonicaj reviewed (commented):
Nice work ;)

I have following concerns:
1) If an application has multiple versions UI shows the latest version of the application by default which is good. However if latest version if not healthy, we should not show it and instead opt to choose the one which is healthy and is supported on the system.

![Screenshot 2021-04-01 at 11 58 36 AM](https://user-images.githubusercontent.com/17968138/113255856-c8420980-92e1-11eb-9beb-c8ef28d69d8c.png)

2) We allow clicking on a catalog row to see the summary of the catalog in question which is nice but we don't change the shape of the cursor which does not make it intuitive for the user so that he/she knows that the row is clickable and an action will happen.

3) We should have `Force Create` as boolean `false` when adding a catalog as a default.

4) Catalogs which have `builtin` as `true` cannot be deleted and we should not show `Delete` option for them.

5) Pulling a private repo works with api but not in the UI ( UI is not sending docker authentication parameters to the api )
```
❯ midclt call -job container.image.pull '{"from_image":"sonicaj/priv", "tag":"test", "docker_authentication": {"username": "sonicaj", "password": "passhere"}}'
```

![Screenshot 2021-04-01 at 12 26 32 PM](https://user-images.githubusercontent.com/17968138/113258538-8024e600-92e5-11eb-8fb9-d3397e4a33eb.png)

6) We should allow user to see what password has been input similar to other password fields we have in the UI ( this is when pulling docker image )

![Screenshot 2021-04-01 at 12 26 45 PM](https://user-images.githubusercontent.com/17968138/113258570-89ae4e00-92e5-11eb-8a75-b881089339f1.png)

Ornias1993 commented on src/app/pages/applications/forms/catalog-add-form.component.ts:
```diff
@@ -0,0 +1,74 @@
+          required: true,
+        },
+        {
+          type: 'checkbox',
+          name: 'force',
+          placeholder: helptext.catalogForm.forceCreate.placeholder,
+          tooltip: helptext.catalogForm.forceCreate.tooltip,
+          value: true,
```
```suggestion
          type: 'checkbox',
          name: 'force',
          placeholder: helptext.catalogForm.forceCreate.placeholder,
          tooltip: helptext.catalogForm.forceCreate.tooltip,
          value: false,
```

easy editable version of the feedback from @sonicaj

AlexKarpov98 commented:
@sonicaj @Ornias1993 Fixed all issues. Please check them.

Ornias1993 commented:
@AlexKarpov98 Everything seems to work nicely, besides one little nitpick:
- We might want to hide buildin containers from the container overview

kmoore134 commented:
Might be handy to have a "Remove All Unused" button eventually as well, but that may need to be an enhancement at a later time.

sonicaj commented:
> @AlexKarpov98 Everything seems to work nicely, besides one little nitpick:
> 
> * We might want to hide buildin containers from the container overview
> 
> As i'm a bit scared of users just blindly updating all "update available" containers and accidantally breaking something.

@Ornias1993 latest nightlies tomorrow should hide all system images. If not, please let us know. Also i assume you are referring to "images" ?

AlexKarpov98 commented:
@sonicaj @Ornias1993 Fixed `markdown` issue. Please check it.
![image](https://user-images.githubusercontent.com/22980553/113338045-c82e2200-9352-11eb-886e-413c828c543b.png)

Ornias1993 commented:
@AlexKarpov98 Basic MDseems to render fine... no idea about things like titles though.
(and small note: i had to hard-reload the browser window after downloading the new docker image, in case someone doesn't get the new text)

sonicaj reviewed (commented):
I have just one concern :p 

1) I think in latest changes we regressed. This catalog item as a whole should not be considered `Healthy` if it has some versions which are not healthy.

![Screenshot 2021-04-01 at 11 04 43 PM](https://user-images.githubusercontent.com/17968138/113335641-a7f46800-933e-11eb-8ae7-1c5f481e36ee.png)

I think it would be nice to have Healthy status here as well.
![Screenshot 2021-04-02 at 12 52 12 AM](https://user-images.githubusercontent.com/17968138/113346730-abdbb680-934d-11eb-9717-4e4a873aa4bd.png)

Ornias1993 commented:
Updated to latest nightly...
No update warnings anymore for system images, which is great.

But some system images still show, along with smo image rows without names:

![image](https://user-images.githubusercontent.com/7613738/113347033-e2590780-9334-11eb-8cc9-522a6228087d.png)

sonicaj commented:
@Ornias1993 i assume you are referring to zfs-ci and csi-snapshotter one, we are not consuming them anymore. Which brings a different question of cleaning up unused system images over time as we move tags. ( good point though :) )

Ornias1993 commented:
@sonicaj Ahh okey, that solves/explains one of the two problems :)

sonicaj commented:
@Ornias1993 can you perhaps DM me output of `midclt call container.image.query | jq .` please ?

Ornias1993 commented:
@sonicaj Done, you got a DM on Slack :)

AlexKarpov98 commented:
> I have just one concern :p
> 
> 1. I think in latest changes we regressed. This catalog item as a whole should not be considered `Healthy` if it has some versions which are not healthy.
> 
> ![Screenshot 2021-04-01 at 11 04 43 PM](https://user-images.githubusercontent.com/17968138/113335641-a7f46800-933e-11eb-8ae7-1c5f481e36ee.png)
> 
> I think it would be nice to have Healthy status here as well.
> ![Screenshot 2021-04-02 at 12 52 12 AM](https://user-images.githubusercontent.com/17968138/113346730-abdbb680-934d-11eb-9717-4e4a873aa4bd.png)

@sonicaj Reverted some changes. Please check again.

sonicaj commented:
@Ornias1993 i have not received the output so far on slack.

sonicaj commented:
@AlexKarpov98 i am still seeing same behaviour. Please see `minio` app for the catalog we have been using for testing.

Ornias1993 commented:
> @Ornias1993 i have not received the output so far on slack.

SHOOT I hit a max size, send again

AlexKarpov98 commented:
@sonicaj Please help me check if this is correct.
![image](https://user-images.githubusercontent.com/22980553/113351109-958d2500-9364-11eb-9829-b7d643e821c9.png)

sonicaj commented:
@AlexKarpov98 i think i tried to pull new docker image too quickly
```
❯ docker rmi -f ixsystems/truenas-webui:5269
Untagged: ixsystems/truenas-webui:5269
Untagged: ixsystems/truenas-webui@sha256:6316ef8b70ac93240c08e3003f3a87582ed38299a7b4927cf6b52c2a192eb5b4
Deleted: sha256:791bd2924c7f636ecac03f9fb611fb64c746ca8306fa9b84162a4ca26dfc3098
❯ docker container run -it -e TNIP=ssh.sonicaj.com -p 8080:80 ixsystems/truenas-webui:5269
Unable to find image 'ixsystems/truenas-webui:5269' locally
5269: Pulling from ixsystems/truenas-webui
75646c2fb410: Already exists
6770d38fee91: Already exists
9be5fee3ba64: Already exists
eb64edd14c27: Already exists
259a7c6f3c0a: Already exists
Digest: sha256:6316ef8b70ac93240c08e3003f3a87582ed38299a7b4927cf6b52c2a192eb5b4
Status: Downloaded newer image for ixsystems/truenas-webui:5269
To connect to WebUI: http://localhost:<port>
```

Trying again, btw did we add the healthy tag/status on catalog summary as well ?

AlexKarpov98 commented:
> @AlexKarpov98 i think i tried to pull new docker image too quickly
> 
> Trying again, btw did we add the healthy tag/status on catalog summary as well ?
@sonicaj Yes, we did, like this
![image](https://user-images.githubusercontent.com/22980553/113351563-4398cf00-9365-11eb-98b6-b1bf0900a555.png)

sonicaj commented:
@AlexKarpov98 i meant showing it above like we do for Application Summary on top right corner. A good use case a catalog which is misconfigured - maybe incorrect URI, we wouldn't have any item/train to show for it and can't tell if it's healthy or not.

sonicaj reviewed (approved):
Really nice work! Thanks

Ornias1993 reviewed (approved):
@AlexKarpov98 

I think you can remove the WIP tag :)
LGTM!

dszidi commented on src/app/pages/common/entity/entity-table/entity-table.component.html:
```diff
@@ -176,7 +176,7 @@
         ix-auto ix-auto-type="expander" ix-auto-identifier="{{element[getFirstKey(element)]}}"
         *matRowDef="let element; columns: columnsToString(currentColumns, 'prop'); let i = dataIndex;"
-        (click)="expandedElement = expandedElement === element ? null : element;"
+        (click)="conf.onRowClick?conf.onRowClick(element):(expandedElement = expandedElement === element ? null : element);"
```
It would be more readable if this was implemented as a method in the .ts file.

dszidi commented:
Placeholder text should be visible even when the cursor is not in the input field

![image](https://user-images.githubusercontent.com/29865743/113441949-2132a000-93bd-11eb-949b-c6295c4205d7.png)

![image](https://user-images.githubusercontent.com/29865743/113441988-37d8f700-93bd-11eb-997a-bc4777b1d898.png)

dszidi commented on src/app/core/components/directives/html-tooltip/html-tooltip.directive.css:
```diff
@@ -0,0 +1,16 @@
+div.html-tooltip{
+  /*position:relative;
```
Let's delete these unused styles

AlexKarpov98 commented:
@dszidi Fixed the issues. Please check them.

dszidi reviewed (approved):
LGTM

Ornias1993 commented:
Found a small bug:
The bulk-selection checkbox on the installed applications overview, isn't properly exluded from the on-click application-details pop-up modal.
(so it pops up the info modal at every selection)
